linode/manager
# refactor: [M3-8228] - Refactor `SupportTicketDialog` to use React Hook Form

## Description 📝
SupportTicketDialog and its subcomponents for custom ticket types (currently SMTP, but Limits to come) are not very extensible. They are not using any form library, but relying on state for managing the form. The file is also already long and growing, at around 800 lines. Finally, there are known issues with the way that the dialog repopulates fields stored in local storage with stale data. Refactoring this component now will make it easier to manage in the future as we add additional custom ticket types.

> [!NOTE]
> - This PR is large, but mostly because of moving variables and components around. Hopefully the self review helps with readability.
> - `scrollIntoView` was not working correctly before the refactor and `scrollErrorIntoViewV2` is not working correctly now. (See [comment](https://github.com/linode/manager/pull/10557#discussion_r1643597590).) Unless reviewers have any suggestions, I'm going to create a ticket to follow up on this in a subsequent PR. 


## Changes  🔄
- Transitioned `SupportTicketDialog` to React Hook Form.
- Moved the fields related to entity type into their own component (`SupportTicketProductSelectionField.tsx`)
- Moved constants to `/SupportTickets/constants.ts` and util to `/SupportTickets/ticketUtils.ts`.
- Fixes a few 🐛 as noted in the self review comments.
- Improved unit (`ticketUtils.test.ts`) and integration (`open-support-ticket.spec.ts`) test coverage.

## Preview 📷

There should be no visual changes except the enabled primary action button when required fields aren't filled out for feedback with validation.

## How to test 🧪

### Prerequisites
(How to setup test environment)
- **Please test this in the dev environment so that we don't bother Customer Support with tickets.**

### Verification steps
(How to verify changes)
- Check out this PR and switch to the dev environment.
- To test a general support ticket:
   - Go to http://localhost:3000/support/tickets
   - Click 'Open New Ticket'. 
   - Click the submit button without filling out any fields and confirm that validation messages display for the general support ticket. 
   - Fill out the ticket. Include an entity like a Linode and confirm that you can select from your available linodes (or none if you have none of those entities).
   - Attach a file. Confirm no regressions.
   - Submit the ticket, and confirm that it is visible in the landing page table of open support tickets.
- To test a SMTP support ticket:
   - Go to the details page of a linode on your account that you haven't dismissed the SMTP warning notice for. If you don't see the notice, create a new linode for this purpose.
   - Click on the text "open a support ticket"in the warning notice.
   - Confirm that you are redirected to the support landing page with an SMTP-specific ticket open.
   - Confirm that some fields are prepopulated always (summary) or if the information exists for the account (customer name, company name).
   - Click the submit button without filling out any additional fields and confirm that validation messages display for the SMTP support ticket. 
   - Fill out the ticket, submit the ticket, and click on the new SMTP ticket in the ticket table. Confirm that the description field includes all the required answers for the custom SMTP-related form fields and it's formatted in Markdown as expected.
- In general: test behavior like closing out of the form (clicking somewhere outside the dialog on the screen) vs. canceling the form. Closing out of the form should save any title or description. Canceling should clear the whole form. Confirm that you can cancel out of a SMTP ticket, reopen the dialog to create a general support ticket, and see the correct form (not SMTP).
- Test that you can still see the ticket severity select and make a selection with the feature flag on.
- Confirm test coverage passes for new tests.

## As an Author I have considered 🤔

*Check all that apply*
